src/plugins/validation/2and3/semantic-validators/items-required-for-array-objects.js

```diff
@@ -36,14 +52,44 @@ module.exports.validate = function({ jsSpec }) {
       }
 
       // Assertation 2
+      const checkRequiredProps = config.operations.check_required_properties;
       if (Array.isArray(obj.required)) {
         obj.required.forEach((requiredProp, i) => {
           if (!obj.properties || !obj.properties[requiredProp]) {
-            messages.addMessage(
-              path.concat([`required[${i}]`]).join('.'),
-              "Schema properties specified as 'required' must be defined",
-              'error'
-            );
+            // For allOf, ensure atleast one of the child objects has the required property
+            if (Array.isArray(obj.allOf)) {
+              var props = {};
+              obj.allOf.forEach((childObj) => {
+                childObj = reduceObj(jsSpec, childObj);
+                props = Object.assign(props, childObj.properties);
+              });
+              if (!props[requiredProp]) {
+                messages.addMessage(
+                  path.concat([`required[${i}]`]).join('.'),
+                  "Schema properties specified as 'required' should be defined",
+                  checkRequiredProps
+                );
+              }
+            // For anyOf/oneOf, ensure all of the child objects have the required property
+            } else if (Array.isArray(obj.anyOf) || Array.isArray(obj.oneOf)) {
+              var childList = obj.anyOf || obj.oneOf;
+              childList.forEach((childObj) => {
+                childObj = reduceObj(jsSpec, childObj);
+                if (childObj.properties && !childObj.properties[requiredProp]) {
+                  messages.addMessage(
+                    path.concat([`required[${i}]`]).join('.'),
+                    "Schema properties specified as 'required' should be defined",
```

Comment:
My concern here is that, if an API definition defines a `oneOf` with three schemas, and two of them are missing the definition, they will get two identical messages without distinguishing which ones are problematic.

Perhaps we should use a message like "Schema should define required property <property name>" and point the path at the schema itself